keps/582-preempt-based-on-flavor-order/README.md

```diff
@@ -0,0 +1,419 @@
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As a Kueue administrator I want to ensure more important jobs running on more 
+stable resources. This can happen in case that there are normal and spot instances
+in my cluster. In this case I prefer my high priority jobs not running on spot 
+instances. If high priority jobs can preempt jobs in standard instances before trying spot instances,
+stability can be achieved.
+
+My use case can be supported by setting `flavorFungibility` to `BeforeNextFlavor`  in the Kueue configuration.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Cluster Queue API
+
+We extend the Cluster Queue API to introduce the new fields: flavorFungibility to opt-in and configure the new behavior.
+
+For each type of resource in each podSet, Kueue will traverse all resource groups and resource flavors to find a available flavor in present. When there are insufficient resources in the flavor, kueue will prioritize preemption or borrowing based on the configured policy. 
+
+```
+const (
+	Borrow FlavorFungibilityPolicy = "Borrow"
+	Preempt  FlavorFungibilityPolicy = "Preempt"
+  TryNextFlavor FlavorFungibilityPolicy = "TryNextFlavor"
+)
+
+type FlavorFungibility struct {
+  // +kubebuilder:validation:Enum="Borrow,TryNextFlavor"
+  WhenCanBorrow FlavorFungibilityPolicy  `json:"whenCanBorrow"`
+  // +kubebuilder:validation:Enum="Preempt,TryNextFlavor"
+  WhenCanPreempt FlavorFungibilityPolicy `json:"whenCanPreempt"`
+}
+
+// ClusterQueuePreemption contains policies to preempt Workloads from this
+// ClusterQueue or the ClusterQueue's cohort.
+type ClusterQueuePreemption struct {
+	...
+	FlavorFungibility FlavorFungibility `json:"flavorFungibility"`
+}
+```
+
+If flavorFungibility is nil in configuration, we will set the `WhenCanBorrow` to `true` and set `WhenCanPreempt` to `false` to maintain consistency with the current behavior.
+
```

Comment:
Explain what's the default policy, which should match the current behavior.

IIUC, the current behavior is:

```yaml
whenCanBorrow: Borrow
whenCanPreempt: TryNextFlavor
```